pennylane/interfaces/tf.py

```diff
@@ -146,21 +147,10 @@ def _TFQNode(*input_, **input_kwargs):
 
         def grad(grad_output, **tfkwargs):
             """Returns the vector-Jacobian product"""
-            diff_indices = None
-            non_diff_indices = set()
-
-            # determine the QNode variables which should be differentiated
-            for differentiable, arg_variable in zip(requires_grad, qnode.arg_vars):
-                if not differentiable:
-                    indices = [i.idx for i in _flatten(arg_variable)]
-                    non_diff_indices.update(indices)
-
-            if non_diff_indices:
-                diff_indices = set(range(qnode.num_variables)) - non_diff_indices
-
             # evaluate the Jacobian matrix of the QNode
             variables = tfkwargs.get("variables", None)
-            jacobian = qnode.jacobian(args, kwargs, wrt=diff_indices)
+            qnode.set_trainable_args({i for i, val in enumerate(requires_grad) if val})
```

Comment:
since `requires_grad` is already being pulled in via function closure, do we need to set the trainable args again? (also seems to be handled by closure)

Reply:
Strangely yes. I thought along the same lines as you, and didn't include it here at first. However, for some reason closure 'wasn't being obeyed' by TF in the backwards pass, and `qnode.trainable_args` would always be `set()` (even though it was non-empty in the forward pass).

I'm still not 100% sure _why_ closure doesn't work here, it must be either a Python gotcha or a TF gotcha 🤔